Unify CWalletTx construction

ryanofsky opened the pull request:
Two commits:

- `Construct CWalletTx objects in CommitTransaction` moves a bunch of CWalletTx initialization into CWallet::CommitTransaction to dedup some code and avoid future inconsistencies in how wallet transactions are created.
- `Get rid of CWalletTx default constructor` does what is described and eliminates the possibility of empty transaction entries being inadvertently created by mapWallet[hash] accesses.

Both of these changes were originally part of #9381

ryanofsky committed:
[wallet] Get rid of CWalletTx default constructor

No change in behavior in the normal case. But buggy mapWallet lookups with
invalid txids will now throw exceptions instead of inserting dummy entries into
the map, and potentially causing segfaults and other failures.

This also makes it a compiler error to use the mapWallet[hash] syntax which
could create dummy entries.

ryanofsky committed:
[wallet] Construct CWalletTx objects in CommitTransaction

Construct CWalletTx objects in CWallet::CommitTransaction, instead of having
callers do it. This ensures CWalletTx objects are constructed in a uniform way
and all fields are set.

This also makes it possible to avoid confusing and wasteful CWalletTx copies in
https://github.com/bitcoin/bitcoin/pull/9381

There is no change in behavior.

ryanofsky commented:
Rebased 84808324df503cf96dd7d3123683a112a3f51815 -> c472c5abd9467de67a20ad814a7c3d1543f6c794 ([pr/makewtx.0](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.0) -> [pr/makewtx.1](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.1)) because of conflicts with pwalletMain renames in #9775.

promag commented on src/qt/walletmodel.cpp:
```diff
@@ -578,9 +578,9 @@ void WalletModel::getOutputs(const std::vector<COutPoint>& vOutpoints, std::vect
         if (!wallet->mapWallet.count(outpoint.hash)) continue;
-        int nDepth = wallet->mapWallet[outpoint.hash].GetDepthInMainChain();
+        int nDepth = wallet->mapWallet.at(outpoint.hash).GetDepthInMainChain();
         if (nDepth < 0) continue;
```
Isn't this condition taking advantage of the dummy `CWalletTx`? If so replace `.at()` with `.find()` and test the iterator.

ryanofsky replied:
> Isn't this condition taking advantage of the dummy CWalletTx? If so replace .at() with .find() and test the iterator.

Dummy constructor isn't called here because of the count check above.

promag commented on src/qt/walletmodel.cpp:
```diff
@@ -578,9 +578,9 @@ void WalletModel::getOutputs(const std::vector<COutPoint>& vOutpoints, std::vect
+        int nDepth = wallet->mapWallet.at(outpoint.hash).GetDepthInMainChain();
         if (nDepth < 0) continue;
-        COutput out(&wallet->mapWallet[outpoint.hash], outpoint.n, nDepth, true /* spendable */, true /* solvable */, true /* safe */);
+        COutput out(&wallet->mapWallet.at(outpoint.hash), outpoint.n, nDepth, true /* spendable */, true /* solvable */, true /* safe */);
```
Another lookup? With the above suggestion it can be avoided.

ryanofsky replied:
> Another lookup? With the above suggestion it can be avoided.

It looks like this is done in your PR #11039. I'll review that PR, so this PR can be limited to replacing [] lookups with at() lookups.

promag commented on src/wallet/feebumper.cpp:
```diff
@@ -246,7 +246,7 @@ bool CFeeBumper::commit(CWallet *pWallet)
         return false;
     }
-    CWalletTx& oldWtx = pWallet->mapWallet[txid];
+    CWalletTx& oldWtx = pWallet->mapWallet.at(txid);
```
Replace with `.find(txid)` above (before `.count()`) and test if exists, also avoids 2nd lookup.

ryanofsky replied:
> Replace with .find(txid) above (before .count()) and test if exists, also avoids 2nd lookup.

Also seems to be taken care of in #11039.

promag reviewed (commented):
`mapWallet.at()` can throw so I guess it should be avoided and replaced with `mapWallet.find()`, to test the returned iterator.

Needs rebase.

ryanofsky reviewed (commented):
Thanks very much for the review. Rebased 2245b37ea967ca0047766479cd8e85409399fe0a -> 51b93a90a403abed3716782d86dada3d4e92e38e ([pr/makewtx.4](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.4) -> [pr/makewtx.5](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.5)), just fixing minor merge conflicts.

laanwj commented:
> Get rid of CWalletTx default constructor does what is described and eliminates the possibility of empty transaction entries being inadvertently created by mapWallet[hash] accesses.

Great! I've always been wary of this projects' gratuitous use of `[]` for retrieval, as we don't want to accidentally create empty objects and because it generates more code. The second is true even when it's shielded with `.count()` or such. `.at` is better.

So concept ACK at least - haven't been able to review all the changes in detail.

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -379,7 +379,7 @@ UniValue getaddressesbyaccount(const JSONRPCRequest& request)
 }
 
-static void SendMoney(CWallet * const pwallet, const CTxDestination &address, CAmount nValue, bool fSubtractFeeFromAmount, CWalletTx& wtxNew, const CCoinControl& coin_control)
+static void SendMoney(CWallet * const pwallet, const CTxDestination &address, CAmount nValue, bool fSubtractFeeFromAmount, const CCoinControl& coin_control, mapValue_t mapValue, std::string fromAccount, CTransactionRef& txNew)
```
Given that the argument `txNew` seems to be only used to store the resulting output in, wouldn't it be more convenient to make it the return type rather than an argument?

ryanofsky replied:
> Given that the argument txNew seems to be only used to store the resulting output in, wouldn't it be more convenient to make it the return type rather than an argument?

Forgot to respond to this commit, but this was a good idea, and I implemented it in a previous update to the PR: a0a4d8205498c38e53fb85e25cf6428ead1e93e1 -> f16d28cc4cbab3b1c8d05d9e327676afafefdbdf ([pr/makewtx.11](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.11) -> [pr/makewtx.12](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.12))

jonasschnelli commented:
Nice!
Concept ACK.

sipa commented on src/qt/walletmodel.cpp:
```diff
@@ -320,22 +320,22 @@ WalletModel::SendCoinsReturn WalletModel::sendCoins(WalletModelTransaction &tran
 
         CDataStream ssTx(SER_NETWORK, PROTOCOL_VERSION);
-        ssTx << *newTx->tx;
+        ssTx << **newTx;
```
I think the extra dereference isn't needed here; serializing a `std::shared_ptr<X>` should work, if `X` is serializable.

ryanofsky replied:
Done in d070a69e9734b83e2defc29aa97bf4dab8953151

sipa reviewed (commented):
utACK f16d28cc4cbab3b1c8d05d9e327676afafefdbdf

ryanofsky reviewed (commented):
Added 1 commit f16d28cc4cbab3b1c8d05d9e327676afafefdbdf -> d070a69e9734b83e2defc29aa97bf4dab8953151 ([pr/makewtx.12](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.12) -> [pr/makewtx.13](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/makewtx.12...pr/makewtx.13))
Squashed d070a69e9734b83e2defc29aa97bf4dab8953151 -> 1dc99fb74210e511287e1b2c0406ab336ffa241f ([pr/makewtx.13](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.13) -> [pr/makewtx.14](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.14))
Rebased 1dc99fb74210e511287e1b2c0406ab336ffa241f -> fe84e57ad18f7e5518a191aa30d70fc37ff3f73e ([pr/makewtx.14](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.14) -> [pr/makewtx.15](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.15))
Rebased fe84e57ad18f7e5518a191aa30d70fc37ff3f73e -> 857a50a93fc64c8c7af1d7dd05fc4992dc170c0d ([pr/makewtx.15](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.15) -> [pr/makewtx.16](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.16))
Rebased 857a50a93fc64c8c7af1d7dd05fc4992dc170c0d -> d8f17e6388787da71ca502d74abdfe0240c315c0 ([pr/makewtx.16](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.16) -> [pr/makewtx.17](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.17))
Rebased d8f17e6388787da71ca502d74abdfe0240c315c0 -> 1531d0df4bd727c5c30d4685ab9abeb0810936db ([pr/makewtx.17](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.17) -> [pr/makewtx.18](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.18))
Rebased 1531d0df4bd727c5c30d4685ab9abeb0810936db -> 0c6ee54c94f2972e084022c539ffc52345daf48f ([pr/makewtx.18](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.18) -> [pr/makewtx.19](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.19))
Rebased 0c6ee54c94f2972e084022c539ffc52345daf48f -> b110df7d9548566c428a2449f1e5b92633b98878 ([pr/makewtx.19](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.19) -> [pr/makewtx.20](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.20))
Rebased b110df7d9548566c428a2449f1e5b92633b98878 -> 7bae52b4f5d7893a951c38af032a09b12c4ccebf ([pr/makewtx.20](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.20) -> [pr/makewtx.21](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.21))

jamesob commented on src/wallet/test/accounting_tests.cpp:
```diff
@@ -44,7 +44,7 @@ BOOST_AUTO_TEST_CASE(acc_orderupgrade)
     wtx.mapValue["comment"] = "z";
     pwalletMain->AddToWallet(wtx);
-    vpwtx.push_back(&pwalletMain->mapWallet[wtx.GetHash()]);
+    vpwtx.push_back(&pwalletMain->mapWallet.at(wtx.GetHash()));
```
The commit message seems to suggest `.at()` will cause compile-time errors for non-existent keys, but from what I understand `at()` will throw runtime `out_of_range` exceptions. This change in behavior seems okay, but I don't have great context here -- are we alright with these lines now being liable to throw runtime exceptions?

ryanofsky replied:
Oops, commit message is mistaken, it should say "runtime exceptions" not "compile errors". But the change is intentional because it should be preferable to throw exceptions if invalid txids are looked up instead of creating empty map entries with null CTransactionRef pointers that might later get dereferenced.

jamesob reviewed (commented):
Concept ACK https://github.com/bitcoin/bitcoin/pull/9680/commits/7bae52b4f5d7893a951c38af032a09b12c4ccebf

jamesob commented:
ACK https://github.com/bitcoin/bitcoin/commit/7bae52b4f5d7893a951c38af032a09b12c4ccebf

Cloned this branch, compiled, and ran `bitcoin-qt`.

ryanofsky commented:
Updated commit message 7bae52b4f5d7893a951c38af032a09b12c4ccebf -> a5a2e7eddeaa3c2323da83f1b40b1c382718cc3a ([pr/makewtx.21](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.21) -> [pr/makewtx.22](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.22)), no other changes.

jamesob commented:
ACK https://github.com/bitcoin/bitcoin/commit/a5a2e7eddeaa3c2323da83f1b40b1c382718cc3a

ryanofsky commented:
Rebased a5a2e7eddeaa3c2323da83f1b40b1c382718cc3a -> da94a51fa50766d4bb528a16fee2b6e7a1fb2146 ([pr/makewtx.22](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.22) -> [pr/makewtx.23](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.23)) due to conflict with #12421 causing travis build failures.

MarcoFalke commented:
Needs rebase

ryanofsky commented:
Rebased 8fcd24ece6720d09a22758d6addafcf3539057a2 -> 2c08d0b47843de4aab10e21db7a417791e34be2e ([pr/makewtx.24](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.24) -> [pr/makewtx.25](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.25)) due to conflict with #12607.

ryanofsky commented:
Added 1 commits 2c08d0b47843de4aab10e21db7a417791e34be2e -> d0b85362d3f620cf397b00992fddc48a5fc4ee79 ([pr/makewtx.25](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.25) -> [pr/makewtx.26](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.26), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/makewtx.25...pr/makewtx.26)) implementing suggestion from https://github.com/bitcoin/bitcoin/pull/9381#discussion_r169884033.
Squashed d0b85362d3f620cf397b00992fddc48a5fc4ee79 -> b4bc32a451720167000f59dd73ab07990f9c6b92 ([pr/makewtx.26](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.26) -> [pr/makewtx.27](https://github.com/ryanofsky/bitcoin/commits/pr/makewtx.27))

kallewoof reviewed (approved):
utACK a128bdc9e15dec5cd9aed1e4922c938edf31eb9a

sipa commented:
utACK b4bc32a451720167000f59dd73ab07990f9c6b92
